File: apps/api/TourPlanner.Models/Tour.cs
using TourPlanner.Models.Enums;

namespace TourPlanner.Models;

public class Tour
{
    public Guid Id { get; set; } = Guid.NewGuid();

    public string UserId { get; set; } = string.Empty;

    public string Name { get; set; } = string.Empty;

    public string Description { get; set; } = string.Empty;

    public string From { get; set; } = string.Empty;

    public string To { get; set; } = string.Empty;

    public TransportType TransportType { get; set; }

    public double? Distance { get; set; }

    public double? EstimatedTimeSeconds { get; set; }

    // GeoJSON route geometry from ORS
    public string? RouteInformation { get; set; }

    public string? ImagePath { get; set; }

    public DateTime CreatedAt { get; set; } = DateTime.UtcNow;
    public DateTime UpdatedAt { get; set; } = DateTime.UtcNow;

    public ICollection<TourLog> TourLogs { get; set; } = new List<TourLog>();

    public int Popularity => TourLogs.Count;

    /// <summary>
    /// Child-friendliness derived from recorded difficulty, total times and distance.
    /// VerySuitable: all logs are Easy and distance ≤ 10 km
    /// Suitable:     average difficulty is Easy/Medium and distance ≤ 30 km
    /// NotSuitable:  otherwise
    /// </summary>
    public ChildFriendliness ChildFriendliness
    {
        get
        {
            if (!TourLogs.Any()) return ChildFriendliness.NotSuitable;

            var avgDifficulty = TourLogs.Average(l => (int)l.Difficulty);
            var avgDistance = TourLogs.Average(l => l.TotalDistanceKm);

            if (avgDifficulty <= (int)Difficulty.Easy && avgDistance <= 10)
                return ChildFriendliness.VerySuitable;

            if (avgDifficulty <= (int)Difficulty.Medium && avgDistance <= 30)
                return ChildFriendliness.Suitable;

            return ChildFriendliness.NotSuitable;
        }
    }
}

File: apps/web/src/models/tour.ts
export type TransportType = 'Car' | 'Bicycle' | 'Hiking' | 'Running' | 'Vacation'
export type ChildFriendliness = 'NotSuitable' | 'Suitable' | 'VerySuitable'

export interface Tour {
  id: string
  name: string
  description: string
  from: string
  to: string
  transportType: TransportType
  distance?: number | null
  estimatedTimeSeconds?: number | null
  routeInformation?: string | null
  imagePath?: string | null
  popularity: number
  childFriendliness: ChildFriendliness
  createdAt: string
  updatedAt: string
}

export interface CreateTourPayload {
  name: string
  description: string
  from: string
  to: string
  transportType: TransportType
}

export interface UpdateTourPayload {
  name: string
  description: string
  from: string
  to: string
  transportType: TransportType
  distance?: number | null
  estimatedTimeSeconds?: number | null
  routeInformation?: string | null
}

export interface TourRoute {
  coordinates: [number, number][]
  elevations: number[]
  distanceMeters: number
  durationSeconds: number
}

export function parseRoute(routeInformation?: string | null): TourRoute | null {
  if (!routeInformation) return null
  try {
    const r = JSON.parse(routeInformation)
    if (!Array.isArray(r.coordinates) || r.coordinates.length < 2) return null
    return {
      coordinates: r.coordinates as [number, number][],
      elevations: Array.isArray(r.elevations) ? r.elevations : [],
      distanceMeters: typeof r.distanceMeters === 'number' ? r.distanceMeters : 0,
      durationSeconds: typeof r.durationSeconds === 'number' ? r.durationSeconds : 0,
    }
  } catch {
    return null
  }
}

export const TRANSPORT_TYPES: TransportType[] = ['Car', 'Bicycle', 'Hiking', 'Running', 'Vacation']

export const TRANSPORT_ICONS: Record<TransportType, string> = {
  Car: '🚗',
  Bicycle: '🚴',
  Hiking: '🥾',
  Running: '🏃',
  Vacation: '✈️',
}
